pybamm/expression_tree/operations/latexify.py

```diff
@@ -0,0 +1,381 @@
+#
+# Latexify class
+#
+import copy
+import operator
+import re
+import warnings
+
+import sympy
+
+import pybamm
+from pybamm.expression_tree.printing.sympy_overrides import custom_print_func
+
+
+class Latexify:
+    """
+    Converts all model equations in latex.
+
+    Parameters
+    ----------
+    filename: str (optional)
+        Accepted file formats - any image format, pdf and tex
+        Default is None, When None returns all model equations in latex
+        If not None, returns all model equations in given file format.
+
+    newline: bool (optional)
+        Default is True, If True, returns every equation in a new line.
+        If False, returns the list of all the equations.
+
+    Load model
+    >>> model = pybamm.lithium_ion.SPM()
+
+    This will returns all model equations in png
+    >>> model.latexify("equations.png")
+
+    This will return all the model equations in latex
+    >>> model.latexify()
+
+    This will return the list of all the model equations
+    >>> model.latexify(newline=False)
+
+    This will return first five model equations
+    >>> model.latexify(newline=False)[1:5]
+    """
+
+    def __init__(self, model, filename=None, newline=True):
+        self.model = model
+        self.filename = filename
+        self.newline = newline
+
+    def _get_concat_displays(self, node):
+        """
+        Returns all the concatenation nodes by doing a depth first search through the
+        entire equation tree with ranges in front of all nodes.
+        """
+        concat_displays = []
+        dfs_nodes = [node]
+        while dfs_nodes:
+            node = dfs_nodes.pop()
+            if (
+                hasattr(node, "concat_latex")
+                and getattr(node, "print_name", None) is not None
+            ):
+                # Combine list of concatenations with list of ranges
+                concat_geo = map(
+                    operator.add,
+                    node.concat_latex,
+                    self._get_concat_geometry_displays(node),
+                )
+
+                # Add cases and split by new line
+                concat_sym = (
+                    r"\begin{cases}" + r" \\ ".join(concat_geo) + r"\end{cases}"
+                )
+                concat_eqn = sympy.Eq(
+                    sympy.symbols(node.print_name),
+                    sympy.Symbol(concat_sym),
+                    evaluate=False,
+                )
+                concat_displays.append(concat_eqn)
+            dfs_nodes.extend(node.children)
+
+        # Remove duplicates from the list whilst preserving order
+        return list(dict.fromkeys(concat_displays))
+
+    def _get_concat_geometry_displays(self, var):
+        """Returns a list of min/max ranges of all concatenation nodes in latex."""
+        geo = []
+
+        if not var.domain:
+            return geo
```

Comment:
I don't think these lines are needed, if `var.domain` is empty it will just skip the for loop below

Reply:
Good catch!